File: Trees/Balanced Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int depth(TreeNode* root) {
        if (root == nullptr) return 0;
        return max(depth(root->left), depth(root->right)) + 1;
    }
public:
    bool isBalanced(TreeNode* root) {
        if (root == nullptr) return true;

        int diff = depth(root->left) - depth(root->right);
        return !(diff > 1 || diff < -1) && isBalanced(root->left) && isBalanced(root->right);
    }
};

File: Trees/Balanced Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isBalanced(root *TreeNode) bool {
    if root == nil {
        return true;
    }

    diff := depth(root.Left) - depth(root.Right);
    return !(diff < -1 || diff > 1) && isBalanced(root.Left) && isBalanced(root.Right);
}

func depth(root *TreeNode) int {
    if root == nil {
        return 0;
    }
    return max(depth(root.Left), depth(root.Right)) + 1;
}